patterns.cpp

```diff
@@ -0,0 +1,54 @@
+#include <vector>
```

Comment:
Should separete in hpp and cpp files and add #ifndef #define part, so one could include the file more than once.